### PR TITLE
[Core] Assemble utilities

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This is a utility to assemble nodal values given via a std::unordered_map (or python dict) to nodes. The map is constructed with node id, and value. In the OMP case, it just adds given `nodal_values_map`'s values to respective node.

In the case of MPI, this utility assembles all the `nodal_values_map`'s values accordingly in the model part. The `nodal_values_map` in each rank can hold `node_ids` which are not in their `LocalMesh` or `GhostMesh`.

The same interfaces are extended for `Element` and `Condition` types as well. (Proper tests are also added in OMP and MPI)

**Changelog**
- Added `AssembleUtilities`
- Added OMP and MPI tests

### CODE REVIEW

**Comment by RiccardoRossi** [comment]
@sunethwarna could u expand about what is the goal of this function?

**Comment by sunethwarna** [comment]
Hi @RiccardoRossi,  Its for the following case.

The primal solution consist of `SLIP` flag, where primal residual equation is rotated according to nodal `NORMAL`. If I want to calcualte `SHAPE_SENSITIVITY` w.r.t. nodal coordinates then I need
1. adjoints of this primal problem, there I have to rotate adjoint equation system and solve the adjoint equation system (WIP, and will be in FluidApp). 
2. Afterwards I need to add contributions coming from rotation matrix shape sensitivity via `SensitivityBuilder` (WIP, there will be a `fluid_sensitivity_builder` for this).

This PR is to solve one of the problems I have w.r.t. step 2. Rotation matrix is based on nodal `NORMAL`, and `NORMAL` is calculated from averaging neighbour condition `NORMAL` values. Therefore, shape derivative of `NORMAL` w.r.t. nodal coordinates will have contributions from all the nodal neighbours. I already calculate this proper nodal normal shape sensitivity w.r.t. all of relevant nodes using #7396 -> #7412 -> #7428. Then the problem is, once I have the nodal shape derivative contributions from nodal rotation matrix, some of the nodes' shape sensitivity contribution values doesn't belong to `LocalMesh` or `GhostMesh` nodes. So I need a way to assemble and update nodes' shape sensitivities properly. Therefore I created this utility. :)

I didn't include it in `mpi_communicator` because of the headers I am using in here.

**Comment by RiccardoRossi** [comment]
Ok, but i am still a little lost.

this is designed to be used inside the builder and solver? or when?

how would the input look like?

**Comment by sunethwarna** [comment]
@RiccardoRossi This will be used inside the `SensitivityBuilder`
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/sensitivity_builder.h

The input will look like the following,
In the case of array_1d<double, 3> values (e.g. `Kratos.SHAPE_SENSITIVITY`),
```python
`Process with rank = 0`
'TMap<array_1d<double, 3>>& rNodalValuesMap' = 
            { 1: Kratos.Array3([1.0, 2.0, 3.0]),
              2: Kratos.Array3([3.0, 4.0, 3.0]),
              3: Kratos.Array3([6.0, 8.0, 9.0])
            }

`Process with rank = 1`
'TMap<array_1d<double, 3>>& rNodalValuesMap' = 
            { 2: Kratos.Array3([3.0, 1.0, 1.0])
              1: Kratos.Array3([2.0, 1.0, 1.0]),
            }
```
The above example tells this method to assemble `SHAPE_SENSITIVITY` variable for given nodes by each rank. So at the end, the `SHAPE_SENSITIVITY` variable of the nodes will have following values (assuming `SHAPE_SENSITIVITY` is initialized to `ZeroVector(3)` before calling this method.)
```python
node[1] = [3.0, 3.0, 4.0]
node[2] = [6.0, 5.0, 4.0]
node[3] = [6.0, 8.0, 9.0]
```
All the other nodes will be not changed.

**Comment by RiccardoRossi** [comment]
Imagine that both rank 0 and 1 write to, say, 6

in that case the entry would be summed up and then syncronized between all the nodes?

**Comment by RiccardoRossi** [comment]
> Imagine that both rank 0 and 1 write to, say, 6
> 
> in that case the entry would be summed up and then syncronized between all the nodes?

sorry stupid question, i see that that is happening for 2

File: kratos/mpi/utilities/mpi_assemble_utilities.h [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+                    }
+
+                    KRATOS_ERROR_IF(all_node_id_ranks[Index] == -1)
+                        << "Node id " << all_node_ids[Index]
+                        << " not found in any of the ranks.\n";
+                });
+            }
+
+            // now broadcast all the ranks properly to all ranks
+            r_data_communicator.Broadcast(all_node_id_ranks, 0);
```
Comment (RiccardoRossi) [inline kratos/mpi/utilities/mpi_assemble_utilities.h]: ouch this is very expensive, a all to all communication. I don't have a smarter way of doing this, however it is definitely very expensive.

**Review by RiccardoRossi — COMMENTED** [review]
Did u consiser using global pointers for this?

there is a function to retrieve global pointers to nodes by id, and once you have that the code of this function would become trivial (admittedly the global pointers retrieve fuction does something similar to what you are doing here under the hood, however i think that using that would be better for maintenance)

take a look at 

https://github.com/KratosMultiphysics/Kratos/blob/31c61f3c56f2b2a30a83244b362d4352df444adb/kratos/mpi/tests/sources/test_pointer_communicator.cpp#L50

**Comment by RiccardoRossi** [comment]
one more comment, as it comes to my mind: when you have the id, you also have the node right? if that is so, you know the partition_index and that would allow you to avoid the communication

**Comment by sunethwarna** [comment]
@RiccardoRossi, Thanks for the suggestion, I made use of `GlobalPointers`, now the code is less, and easy to maintain :)

> one more comment, as it comes to my mind: when you have the id, you also have the node right? if that is so, you know the partition_index and that would allow you to avoid the communication

The problem is, I only have the node id, not the node. So I still need to communicate. I am assembling nodal `NORMAL` derivatives according to `nodal_neighbour_map` so I don't have the nodes, I only have the ids. :/ Following is the procedure:
1. First I use #7396 to calculate `Condition` `NORMAL` sensitivities w.r.t. their nodes. (and stored in `Condition`)
2. Then I use #7412 to distribute those `Condition` based `NORMAL` sensitivities to nodes according to given nodal_neighbour_id_map. (stored in non-historical data value container as a `Matrix`)
3. Next these nodal `NORMAL` shape sensitivities are passed in #7428 to get the rotation matrix shape sensitivities. 
4. Afterwards the same nodal_neighbour_id_map is used to map coorectly the `SHAPE_SENSITIVITIES` for each node.

File: kratos/mpi/utilities/mpi_assemble_utilities.h [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+            GlobalPointerCommunicator<NodeType> pointer_comm(
+                r_data_communicator, gp_list.ptr_begin(), gp_list.ptr_end());
+
+            KRATOS_ERROR_IF(!rModelPart.HasNodalSolutionStepVariable(PARTITION_INDEX)) << "PARTITION_INDEX variable is not found in nodal solution step variables list of "
+                                                                                       << rModelPart
+                                                                                              .Name()
+                                                                                       << ".\n";
+
+            auto partition_index_proxy =
```
Comment (RiccardoRossi) [inline kratos/mpi/utilities/mpi_assemble_utilities.h]: you do not need this, the owner_rank is part of the globalpointer

**Review by RiccardoRossi — COMMENTED** [review]
@sunethwarna, 
sorry for delaying you on this, but can we have a short chat by voice about this PR?
I still have the feeling some of the stuff could be simplified a lot ... (but maybe i am wrong)

**Comment by sunethwarna** [comment]
@RiccardoRossi yes for sure. :) I'll msg you.

**Comment by sunethwarna** [comment]
This is no longer required since we now have `GlobalPointerMapCommunicator` from #7826, therefore closing :)

**Pull request closed without merging**